File: coolercontrold/daemon/src/api/stats.rs
// SPDX-FileCopyrightText: 2026 Guy Boldon, Eren Simsek and contributors
// SPDX-License-Identifier: GPL-3.0-or-later

use crate::api::AppState;
use crate::device::{ChannelDataType, ChannelName, ChannelStats, TempName, UID};
use axum::extract::State;
use axum::Json;
use schemars::JsonSchema;
use serde::Serialize;
use std::collections::HashMap;

/// Top-level response for `GET /stats` and `DELETE /stats`. One entry per
/// known device. Channels/temps are only present once they have been
/// observed at least once since daemon start (or last reset).
#[derive(Debug, Clone, Default, Serialize, JsonSchema)]
pub struct StatsResponse {
    pub devices: Vec<DeviceStatsDto>,
}

#[derive(Debug, Clone, Serialize, JsonSchema)]
pub struct DeviceStatsDto {
    pub uid: UID,
    pub temps: HashMap<TempName, ChannelStats>,
    pub channels: HashMap<ChannelName, HashMap<ChannelDataType, ChannelStats>>,
}

pub async fn get_all(State(AppState { stats_handle, .. }): State<AppState>) -> Json<StatsResponse> {
    Json(stats_handle.all().await)
}

pub async fn delete_all(
    State(AppState { stats_handle, .. }): State<AppState>,
) -> Json<StatsResponse> {
    Json(stats_handle.reset_all().await)
}

File: coolercontrol-ui/src/models/Stats.ts
// SPDX-FileCopyrightText: 2022 Guy Boldon, Eren Simsek and contributors
// SPDX-License-Identifier: GPL-3.0-or-later

import type { UID } from '@/models/Device'

export interface ChannelStats {
    min: number
    max: number
    avg: number
    count: number
}

// Wire discriminator from the daemon's ChannelDataType (SCREAMING_SNAKE_CASE).
// Temps are tracked separately under `temps` and so are not represented here.
export type ChannelStatField = 'DUTY' | 'RPM' | 'FREQ' | 'WATTS'

export interface DeviceStatsDTO {
    uid: UID
    temps: Record<string, ChannelStats>
    channels: Record<string, Partial<Record<ChannelStatField, ChannelStats>>>
}

export interface StatsResponseDTO {
    devices: DeviceStatsDTO[]
}

export function defaultStatsResponse(): StatsResponseDTO {
    return { devices: [] }
}
